TAN-2985 Approval flow front end

IvaKop opened the pull request:
For context: https://go-vocal.slack.com/archives/C07Q9Q4FJHE/p1730991943502159
# Changelog
<!-- Replace this comment by a bullet list. More info: https://www.notion.so/citizenlab/Changelog-How-it-works-f418426c75994454a332bf067634f3f1 -->

IvaKop committed:
Rename to project review in the code

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fetch request review

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' of github.com:CitizenLabDotCo/citizenlab into TAN-2985/approval-flow-front-end

IvaKop committed:
Add approval request

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Clean up logic

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' of github.com:CitizenLabDotCo/citizenlab into TAN-2985/approval-flow-front-end

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Clean up logic

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' of github.com:CitizenLabDotCo/citizenlab into TAN-2985/approval-flow-front-end

IvaKop committed:
Fix copy

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Add tests

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' of github.com:CitizenLabDotCo/citizenlab into TAN-2985/approval-flow-front-end

IvaKop committed:
Add first published at attribute

IvaKop committed:
Enable new project button for project managers

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Add commented lines for first published at

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' of github.com:CitizenLabDotCo/citizenlab into TAN-2985/approval-flow-front-end

IvaKop committed:
Fix tooltip

IvaKop committed:
Merge remote-tracking branch 'origin/TAN-3144/first-published-at-attribute' into TAN-2985/approval-flow-front-end

IvaKop committed:
Merge first_published_at attribute

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix build

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' of github.com:CitizenLabDotCo/citizenlab into TAN-2985/approval-flow-front-end

adessy committed:
[TAN-3177] Allow project moderators to delete draft projects

Project moderators are now allowed to delete projects they moderate that have never been published. In other words, if a project was published in the past and then moved back to draft, its moderators are no longer allowed to delete it.

adessy committed:
[TAN-3177] Rework `save_project`

Refactor `WebApi::V1::ProjectsController#save_project` to make it more functional (avoiding the use of the @project instance variable) as preparatory work before enabling moderators to create draft projects.

adessy committed:
[TAN-3177] Adjust moderator permissions for project status updates

PMs are no longer allowed to update the publication status of a project unless it has been approved (via a project review) or it has already been published in the past.

adessy committed:
[TAN-3177] Allow moderators to create draft projects

The default assignee is now set in the `after_create` side effect instead of the `before_create`. This is because the moderator who created the project must first be added as a project moderator for the new project before they can be assigned as the default assignee. This can only be done after the project is saved because the project ID is required to make the user a project moderator.

Additionally, the `after_create` side effect is now included in the transaction to ensure the project is not created if the user cannot be promoted to project moderator.

adessy committed:
[TAN-3177] Refactor WebApi::V1::ProjectsController#create

Removed the unnecessary use of an instance variable to store the project.

IvaKop committed:
Allow project managers to copy and delete projects when appropriate

IvaKop committed:
Fix test

IvaKop committed:
Stabilize e2e test

IvaKop committed:
Merge branch 'master' into TAN-2985/approval-flow-front-end

IvaKop committed:
Add state to project review

Commit:
Translations updated by CI (extract-intl)

adessy committed:
[TAN-3177] Allow moderators to copy projects

IvaKop committed:
Merge branch 'master' into TAN-2985/approval-flow-front-end

IvaKop committed:
Add project filter for pending and approved projects

adessy committed:
[TAN-3177] Fix rubocop offense

IvaKop committed:
Fix build

IvaKop committed:
Make consistent with master

IvaKop committed:
Add upsell trigger

IvaKop committed:
Add confirmation modal for preview link

IvaKop committed:
Add tracks

Commit:
Translations updated by CI (extract-intl)

amanda-anderson commented on front/app/api/project_reviews/useApproveProject.ts:
```diff
@@ -0,0 +1,27 @@
+  return useMutation<ProjectReview, CLErrors, string>({
+    mutationFn: approveProjectReview,
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: projectReviewKeys.items() });
```
Is it possible to only invalidate the specific project item here, rather than all items?

amanda-anderson commented on front/app/api/project_reviews/useRequestProjectReview.ts:
```diff
@@ -0,0 +1,27 @@
+  return useMutation<ProjectReview, CLErrors, string>({
+    mutationFn: requestProjectReview,
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: projectReviewKeys.items() });
```
Same question here - could we invalidate a specific item instead? Or do we need to invalidate all items?

amanda-anderson commented on front/app/containers/Admin/projects/all/index.tsx:
```diff
@@ -188,24 +180,17 @@ const AdminProjectsList = memo(({ className }: Props) => {
+                buttonStyle="admin-dark"
+              >
+                <FormattedMessage {...messages.newProject} />
+              </Button>
```
Just to confirm, this means that any Admin, Folder Mod, or Project Mod can create new projects then right?

IvaKop replied:
Yes, draft projects. Project moderators cannot publish them the first time without approval. Folder moderators and admin can publish projects at any time.

amanda-anderson commented on front/app/containers/Admin/projects/components/ProjectRow/ProjectMoreActionsMenu.tsx:
```diff
@@ -32,21 +33,26 @@ const ProjectMoreActionsMenu = ({
-  const userCanDeleteProject = isAdmin(authUser);
+  const userCanDeleteProject =
+    isAdmin(authUser) || !project.data.attributes.first_published_at;
+
```
Does this mean Folder Mods or Project Mods can't delete draft projects? :thinking:

IvaKop replied:
Admins can delete projects at any time. Folder and project moderators can only delete draft projects that have not been published yet. Otherwise they cannot delete them.

amanda-anderson commented on front/app/containers/Admin/projects/project/projectHeader/ReviewFlow/ReviewRequest.tsx:
```diff
@@ -0,0 +1,53 @@
+  projectId: string;
+}) => {
+  const { formatMessage } = useIntl();
+  const { mutate: requestProjectRevuiew } = useRequestProjectReview();
```
Very minor: Small typo

amanda-anderson reviewed (approved):
A few small questions but nothing blocking! :rocket: 

I didn't do a functional review, so let me know if you need that as well :)

IvaKop committed:
Fix project preview e2e test

IvaKop committed:
Merge branch '3162-preview-link-final-details' of github.com:CitizenLabDotCo/citizenlab into 3162-preview-link-final-details

IvaKop committed:
Address PR comments

IvaKop commented:
Thanks so much for the review, I know it was a big one! 🙌  No need for functional review for now - we will take care of this on our end as there is more to combine into this branch before it works as expected.

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' into TAN-3136/project-approval-tabs

IvaKop committed:
Add tabs to project list

IvaKop committed:
Add spec

IvaKop committed:
Fix blint

IvaKop committed:
Use blank instead of present

IvaKop committed:
Fix rubocop

IvaKop committed:
Fix indentation

IvaKop committed:
Remove unnecessary item from permitted attributes

IvaKop committed:
Fix typo

IvaKop committed:
Merge pull request #9634 from CitizenLabDotCo/TAN-3136/project-approval-tabs

TAN-3163 Project approval tabs

Commit:
Translations updated by CI (extract-intl)

adessy committed:
Merge pull request #9617 from CitizenLabDotCo/TAN-3177/project-review-pm-new-permissions

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' into 3162-preview-link-final-details

IvaKop committed:
Merge pull request #9619 from CitizenLabDotCo/3162-preview-link-final-details

TAN-3162 Preview link final details

IvaKop committed:
Add approved by label

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix types

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' of github.com:CitizenLabDotCo/citizenlab into TAN-2985/approval-flow-front-end

adessy committed:
[TAN-3197] Add notifications for project review requests

adessy committed:
[TAN-3197] Add notifications for project review approvals

adessy committed:
[TAN-3197] Export and reuse the UserLink component

IvaKop committed:
Add approval flow e2e test

IvaKop committed:
Add tracks

IvaKop committed:
Merge branch 'master' into TAN-2985/approval-flow-front-end

IvaKop committed:
Merge branch 'TAN-2985/approval-flow-front-end' into TAN-3161/approval-flow-e2e-test

Commit:
Translations updated by CI (extract-intl)

Commit:
Translations updated by CI (extract-intl)

adessy committed:
[TAN-3197] Add mail campaign for project review requests

adessy committed:
[TAN-3197] Add mail campaign for project review approvals

adessy committed:
[TAN-3197] Fix flaky test

adessy committed:
[TAN-3197] Fix mailer preview filenames

adessy committed:
[TAN-3197] Wrap ActionMailer::Preview subclasses in a transaction

This change makes it easier to write preview mailers that rely on the database. Any new records or database changes made will now be rolled back after the preview is rendered.

IvaKop committed:
Tweak buttons

IvaKop committed:
Add tooltips

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Remove approval tab

IvaKop committed:
Merge branch 'TAN-3299/approval-flow-ux-ui-tweaks' of github.com:CitizenLabDotCo/citizenlab into TAN-3299/approval-flow-ux-ui-tweaks

adessy committed:
Merge pull request #9670 from CitizenLabDotCo/TAN-3197/project-review-notifications

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Remove comment

IvaKop committed:
Merge pull request #9726 from CitizenLabDotCo/TAN-3299/approval-flow-ux-ui-tweaks

TAN-3299 Approval flow UX / UI tweaks

IvaKop committed:
Delete moderator user after approval flow e2e test

IvaKop committed:
Merge base

IvaKop committed:
Merge pull request #9680 from CitizenLabDotCo/TAN-3161/approval-flow-e2e-test

TAN-3161 Approval flow e2e test

adessy committed:
[TAN-3197] Fix foreign key error when deleting project reviews

adessy committed:
[TAN-3196] Add project-reviews activities to the management feed

adessy committed:
Merge branch 'master' into TAN-2985/approval-flow-front-end

# Conflicts:
#	back/db/structure.sql
#	front/app/translations/admin/en.json
#	front/app/translations/admin/fr-BE.json
#	front/app/translations/admin/fr-FR.json

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
[TAN-3196] Improve code robustness against missing translations

From PR review comments

Co-authored-by: Iva <iva@citizenlab.co>
Co-authored-by: Adrien Dessy <adrien@citizenlab.co>

adessy committed:
Merge pull request #9760 from CitizenLabDotCo/TAN-3196/project-reviews-management-feed

Commit:
Translations updated by CI (extract-intl)

adessy committed:
[TAN-2985] Fix failing tests
